grype/matcher/golang/matcher.go

```diff
@@ -60,6 +61,13 @@ func (m *Matcher) Match(store vulnerability.Provider, p pkg.Package) ([]match.Ma
 		return matches, nil, nil
 	}
 
+	// go1.24 started to generate versions in the form +incompatible+dirty, which
+	// fail to be matched: https://github.com/anchore/grype/issues/2482.
+	if p.Name == mainModule && strings.HasSuffix(p.Version, "+incompatible+dirty") {
```

Comment:
Instead of adding this change here to update the package's version, I think it would be better to update the creation function to properly account for this: https://github.com/anchore/grype/blob/main/grype/version/golang_version.go#L47

Reply:
https://semver.org/#spec-item-10 as per the semver spec all versions should be treated identically irrespective of the build metadata after the '+', thus imho it all should be stripped.

> Build metadata MUST be ignored when determining version precedence.

All of `+.*` should be trimmed from all semver, at syft cataloger level.

To also support any future additions.

Also golang violates the semver spec, as it should be "+incompatible.dirty" as per spec.

Reply:
Right -- so in the `newGolangVersion` function, we're taking action to correct go version strings by trimming `go` prefix, for example. It seems like the right place to look for and trim `+dirty`. Why would it be better to happen in this function?

Reply:
@kzantow i wonder how to instrument for such combo to be used for lots of things. Because yeah, it is unlikely that go had main module specific logic at all.

So it would need to be local checkout of a project, use replaces, point to a directory, which is checked out to a tag, with a dirty tree. But I can't remember if this will generate dependencies versions to be +incompatible+dirty or if it will have no module version information at all.

Also it seems golang is just generating versions in violation to the semver spec. Since it should be dot-separate fields as in "+incompatible.dirty". I opened https://github.com/golang/go/issues/71971 about that.

Reply:
@kzantow @luhring please see above implemented in https://github.com/anchore/grype/pull/2486